rust-runtime/aws-smithy-http/src/body.rs

```diff
@@ -183,14 +188,34 @@ impl SdkBody {
     }
 
     /// Given a function to modify an `SdkBody`, run that function against this `SdkBody` before
-    /// returning the result.
+    /// returning the result. An `SdkBody` returned from this function is not signable. If you
+    /// want to wrap a body and still have it be signable, see [`SdkBody::map_readonly_body`].
     pub fn map(self, f: impl Fn(SdkBody) -> SdkBody + Sync + Send + 'static) -> SdkBody {
         if self.rebuild.is_some() {
             SdkBody::retryable(move || f(self.try_clone().unwrap()))
         } else {
             f(self)
         }
     }
+
+    /// Given a function to modify an `SdkBody`, run that function against this `SdkBody` before
+    /// returning the result. **This function MUST NOT alter the contents of the body.**
+    /// An `SdkBody` returned from this function will be signable so long as the inner
+    /// `SdkBody` is signable. If altering the contents of the inner body is required,
+    /// see [`SdkBody::map`].
+    pub fn map_readonly_body(
+        mut self,
+        f: impl Fn(SdkBody) -> SdkBody + Sync + Send + 'static,
+    ) -> SdkBody {
+        let contents = self.bytes_contents.take();
+        let mut out = if self.rebuild.is_some() {
+            SdkBody::retryable(move || f(self.try_clone().unwrap()))
+        } else {
+            f(self)
+        };
```

Comment:
hmmm I'm not positive this is actually correct. I think this means that in `f` `.bytes()` will return `None`